File: C++/minimum-time-to-reach-target-with-limited-power.cpp
// Time:  O((n * p + e * p) * log(n * p))
// Space: O(n * p)

// dijkstra's algorithm
class Solution {
public:
    vector<long long> minTimeMaxPower(int n, vector<vector<int>>& edges, int power, vector<int>& cost, int source, int target) {
        static const auto& INF = numeric_limits<int64_t>::max();

        vector<vector<pair<int, int>>> adj(n);
        for (const auto& x : edges) {
            adj[x[0]].emplace_back(x[1], x[2]);
        }
        vector<vector<int64_t>> dist(n, vector<int64_t>(power + 1, INF));
        dist[source][power] = 0;
        using T = tuple<int64_t, int, int>;
        priority_queue<T, vector<T>, greater<T>> min_heap;
        min_heap.emplace(dist[source][power], -power, source);
        while (!empty(min_heap)) {
            auto [t, p, u] = min_heap.top(); min_heap.pop();
            p = -p;
            if (t > dist[u][p]) {
                continue;
            }
            if (u == target) {
                return {t, p};
            }
            const auto& np = p - cost[u];
            if (np < 0) {
                continue;
            }
            for (const auto& [v, w] : adj[u]) {
                if (t + w >= dist[v][np]) {
                    continue;
                }
                dist[v][np] = t + w;
                min_heap.emplace(dist[v][np], -np, v);
            }
        }
        return {-1, -1};
    }
};

File: Python/minimum-time-to-reach-target-with-limited-power.py
# Time:  O((n * p + e * p) * log(n * p))
# Space: O(n * p)

import heapq


# dijkstra's algorithm
class Solution(object):
    def minTimeMaxPower(self, n, edges, power, cost, source, target):
        """
        :type n: int
        :type edges: List[List[int]]
        :type power: int
        :type cost: List[int]
        :type source: int
        :type target: int
        :rtype: List[int]
        """
        adj = [[] for _ in xrange(n)]
        for u, v, w in edges:
            adj[u].append((v, w))
        dist = [[float("inf")]*(power+1) for _ in xrange(n)]
        dist[source][power] = 0
        min_heap = [(dist[source][power] , -power, source)]
        while min_heap:
            t, p, u = heapq.heappop(min_heap)
            p = -p
            if t > dist[u][p]:
                continue
            if u == target:
                return [t, p]
            np = p-cost[u]
            if np < 0:
                continue
            for v, w in adj[u]:
                if t+w >= dist[v][np]:
                    continue
                dist[v][np] = t+w
                heapq.heappush(min_heap, (dist[v][np], -np, v))
        return [-1, -1]
